File: Leetcode/Number of Music Playlists/Number of Music Playlist.py
#Problem:https://leetcode.com/problems/number-of-music-playlists/description/
from collections import defaultdict

#Approach:
# Create a mapping from the number of different songs used to the count of playlists.
# Initially there is one playlist with no songs.
# For each additional song, add playlists with every new song, If we have used more than K songs, add playlists with
# all songs not in the most recent K.

class Solution:
    def numMusicPlaylists(self, N, L, K):
        
        used_count = {0: 1}     # used_count[i] is the number of playlists that have used i different songs

        for song in range(L):

            new_used_count = defaultdict(int)
            for used, count in used_count.items():

                new_used_count[used + 1] += count * (N - used)  # add any of the (N - used) unused songs
                if used > K:
                    new_used_count[used] += count * (used - K)  # add any used song not in the most recent K songs

            used_count = new_used_count

        
        return used_count[N] % (10 ** 9 + 7)

File: Leetcode/Number of Music Playlists/Number of Music Playlist.c
/*Problem:https://leetcode.com/problems/number-of-music-playlists/description/

Approach:
Let dp[i][j] be the number of playlists of length i that have exactly j unique songs
Consider dp[i][j]. Last song, we either played a song for the first time or we didn't.
If we did, then we had dp[i-1][j-1] * (N-j) ways to choose it.
If we didn't, then we repeated a previous song in dp[i-1][j] * max(j-K, 0) ways
*/

int numMusicPlaylists(int N, int L, int K)
{
	static const int MOD = 1e9 + 7;
	long dp[L + 1][N + 1];
	memset(dp, 0, sizeof(dp));
	dp[0][0] = 1;
	for (int i = 0; i < L; ++i)
	{
		for (int j = 0; j < N; ++j)
		{
			dp[i + 1][j + 1] = dp[i][j] * (N - j) % MOD;
			if (j >= K)
				dp[i + 1][j + 1] = (dp[i + 1][j + 1] + dp[i][j + 1] * (j + 1 - K)) % MOD;
		}
	}
	return dp[L][N];
}
